modules/storage/src/service/fs.rs

```diff
@@ -149,8 +150,30 @@ impl StorageBackend for FileSystemBackend {
 
     async fn retrieve(
         self,
-        hash: String,
+        hash_key: HashKey,
     ) -> Result<Option<impl Stream<Item = Result<Bytes, Self::Error>>>, Self::Error> {
+        let hash = match hash_key {
+            HashKey::Sha256(inner) => inner,
+            HashKey::Sha384(inner) => {
+                return Err(std::io::Error::new(
+                    std::io::ErrorKind::Unsupported,
+                    HashKeyError::UnsupportedAlgorithm(inner.clone()),
+                ));
+            }
+            HashKey::Sha512(inner) => {
+                return Err(std::io::Error::new(
+                    std::io::ErrorKind::Unsupported,
+                    HashKeyError::UnsupportedAlgorithm(inner.clone()),
+                ));
+            }
+            _ => {
+                return Err(std::io::Error::new(
+                    std::io::ErrorKind::InvalidInput,
+                    HashKeyError::UnsupportedAlgorithm("unknown".to_string()),
+                ));
+            }
```

Comment:
I'm not entirely sure how to deal with the non-exhaustiveness of this, but I think the error should contain the algorithm rather than the digest, right? Something like this...

```suggestion
            _ => {
                let algo = String::from(match hash_key.to_string().split_once(':') {
                    Some((v, _)) => v,
                    _ => "unknown",
                });
                return Err(std::io::Error::new(
                    std::io::ErrorKind::Unsupported,
                    HashKeyError::UnsupportedAlgorithm(algo),
                ));
            }
```